solutions/silver/usaco-895.mdx

```diff
@@ -2,234 +2,265 @@
 id: usaco-895
 source: USACO Silver 2019 January
 title: Icy Perimeter
-author: Tanish Tyagi, Brad Ma, Juheon Rhee
+author: Tanish Tyagi, Brad Ma, Juheon Rhee, David Guo
 ---
 
 [Official Analysis (C++)](http://www.usaco.org/current/data/sol_perimeter_silver_jan19.html)
 
+## Explanation
+
+We can BFS to explore each blob, starting from an unvisited '#' cell.
+During BFS, we can calculate the blob’s area by counting the number of '#' cells visited
+and determine its perimeter by counting the number of adjacent cells,
+reducing the count for each shared edge with neighboring '#' cells.
+
+As we traverse the grid, we can initiate a BFS for each unvisited '#' cell to compute the area and perimeter of all blobs.
+We can track the largest area found, and if multiple blobs share the same largest area, we can choose the one with the smallest perimeter.
+
+Finally, we output the largest area and the corresponding smallest perimeter.
+
 ## Implementation
+
 **Time Complexity:** $\mathcal{O}(N^2)$
 
 <LanguageSection>
+<CPPSection>
 
-<JavaSection>
-
-```java
-import java.io.*;
-import java.util.*;
-
-public class IcyPerimeter {
-	public static void main(String[] args) throws IOException {
-		Kattio io = new Kattio("perimeter");
-		n = io.nextInt();
-
-		// we do n + 2 because there is an extra outer ring for flood fill
-		numberedMatrix = new int[n + 2][n + 2];
-
-		for (int x = 1; x <= n; x++) {  // read matrix
-			String line = io.next();
-			for (int y = 1; y <= n; y++) {
-				String character = Character.toString(line.charAt(y - 1));
-				if (character.equals("#")) {
-					numberedMatrix[x][y] = -1;  // mark # as -1
-				}
-			}
-		}
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
 
-		// mark each connected component with a different number
-		int markNumber = 1;
-		for (int x = 1; x <= n; x++) {
-			for (int y = 1; y <= n; y++) {
-				if (numberedMatrix[x][y] == -1) {  // not visited
-					floodfill(x, y, markNumber);
-					markNumber++;
-				}
-			}
-		}
+// Movement directions (up, down, left, right)
+const vector<pair<int, int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
 
-		int[] rowMovement = {0, 1, 0, -1};  // dx dy arrays
-		int[] columnMovement = {1, 0, -1, 0};
-
-		// tracks the perimeter of each connected component
-		int[] perimeters = new int[most];
-
-		for (int x = 0; x < (n + 2); x++) {
-			for (int y = 0; y < (n + 2); y++) {
-				// if part of a connected component
-				if (numberedMatrix[x][y] > 0) {
-					for (int i = 0; i < 4; i++) {  // try all directions
-						// no need to check for out of bounds because of our
-						// outer ring if there is a zero, that means that we can
-						// add one to the perimeter
-						if (numberedMatrix[x + rowMovement[i]][y + columnMovement[i]] ==
-						    0) {
-							perimeters[numberedMatrix[x][y]] += 1;
-						}
-					}
-				}
-			}
-		}
+pair<int, int> bfs(int row, int col, vector<vector<char>> &grid,
```

Comment:
i think a lambda would be a lot cleaner

granted its syntax might be slightly confusing, but i think it's nicer than having to pass in `n`, `grid`, and `visited`